pjmedia/src/pjmedia-codec/lyra.cpp

```diff
@@ -0,0 +1,652 @@
+                                                &lyra_factory.base);
+    if (status != PJ_SUCCESS) {
+        PJ_LOG(2, (THIS_FILE, "Unable to register the codec factory"));
+        goto on_codec_factory_error;
+    }
+
+    lyra_cfg.bit_rate = PJMEDIA_CODEC_LYRA_DEFAULT_BIT_RATE;
+    pj_ansi_strxcpy(lyra_factory.model_path_buf, LYRA_DEFAULT_PATH,
+                    sizeof(lyra_factory.model_path_buf));
+    lyra_cfg.model_path = pj_str(lyra_factory.model_path_buf);
+
+    return PJ_SUCCESS;
+
+on_codec_factory_error:
+    pj_pool_release(lyra_factory.pool);
+    lyra_factory.pool = NULL;
+    return status;
+}
+
+/*
+ * Unregister lyra codec factory from pjmedia endpoint and deinitialize
+ * the lyra codec library.
+ */
+PJ_DEF(pj_status_t) pjmedia_codec_lyra_deinit(void)
+{
+    pj_status_t status;
+    pjmedia_codec_mgr *codec_mgr;
+
+    if (lyra_factory.pool == NULL)
+        return PJ_SUCCESS;
+
+    /* Get the codec manager */
+    codec_mgr = pjmedia_endpt_get_codec_mgr(lyra_factory.endpt);
+    if (!codec_mgr) {
+        PJ_LOG(2, (THIS_FILE, "Unable to get the codec manager"));
+        pj_pool_release(lyra_factory.pool);
+        lyra_factory.pool = NULL;
+        return PJ_EINVALIDOP;
+    }
+
+    /* Unregister the codec factory */
+    status = pjmedia_codec_mgr_unregister_factory(codec_mgr,
+                                                  &lyra_factory.base);
+    if (status != PJ_SUCCESS)
+        PJ_LOG(2, (THIS_FILE, "Unable to unregister the codec factory"));
+
+    /* Release the memory pool */
+    pj_pool_release(lyra_factory.pool);
+    lyra_factory.pool = NULL;
+    lyra_factory.endpt = NULL;
+
+    return status;
+}
+
+/* Prototypes for lyra factory */
+static pj_status_t lyra_test_alloc(pjmedia_codec_factory *factory,
+                                   const pjmedia_codec_info *info)
+{
+    const pj_str_t lyra_tag = { "lyra", 4 };
+    unsigned i;
+
+    PJ_UNUSED_ARG(factory);
+
+    /* Type MUST be audio. */
+    if (info->type != PJMEDIA_TYPE_AUDIO)
+        return PJMEDIA_CODEC_EUNSUP;
+
+    /* Check encoding name. */
+    if (pj_stricmp(&info->encoding_name, &lyra_tag) != 0)
+        return PJMEDIA_CODEC_EUNSUP;
+
+    /* Check clock-rate */
+    for (i = 0; i < PJ_ARRAY_SIZE(lyra_factory.param); ++i) {
+        if (info->clock_rate == lyra_factory.param[i].clock_rate) {
+            return PJ_SUCCESS;
+        }
+    }
+
+    /* Unsupported, or mode is disabled. */
+    return PJMEDIA_CODEC_EUNSUP;
+}
+
+static pj_status_t lyra_default_attr(pjmedia_codec_factory *factory,
+                                     const pjmedia_codec_info *id,
+                                     pjmedia_codec_param *attr)
+{
+    PJ_ASSERT_RETURN(factory == &lyra_factory.base, PJ_EINVAL);
+
+    unsigned idx = 0;
+    pj_bzero(attr, sizeof(pjmedia_codec_param));
+    attr->info.pt = (pj_uint8_t)id->pt;
+    attr->info.channel_cnt = 1;
+
+    idx = attr->info.pt - PJMEDIA_RTP_PT_LYRA_8;

```

Comment:
Note that PT can be rewritten, using it in index calculation here without range check may or may not be harmful (some old codecs may still use it though). Perhaps better avoiding this if there is a more reliable way (e.g: comparing clock rate perhaps)?